### PR TITLE
approved HIPE: agent test suite

### PR DESCRIPTION
**Pull request by dhh1128**
Signed-off-by: Daniel Hardman <daniel.hardman@gmail.com>

### CODE REVIEW

**Comment by devin-fisher** [comment]
[Rendered Text](https://github.com/dhh1128/indy-hipe/blob/agent-test-suite/text/agent-test-suite-interface/README.md)

**Comment by ryanwest6** [comment]
I like that the test suite would be able to support multiple modes and features. However, one thing that could be updated is the references to interface and content RFCs. It looks like this was written before the concept of HIPEs were put into practice. Perhaps updating the vocabulary would make it more consistent.

File: text/agent-test-suite-interface/README.md [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+# Summary
+[summary]: #summary
+
+Defines the interface (behavioral contract, mental model, procedure,
+harness) of test suites that evaluate interoperability of Indy agents.
+
+##### Related RFCs
+
+This is a test suite __interface RFC__. It is intended to specify
```
Comment (ryanwest6) [inline text/agent-test-suite-interface/README.md]: I like that the test suite would be able to support multiple modes and features. However, one thing that could be updated is the references to interface and content RFCs. It looks like this was written before the concept of HIPEs were put into practice. Perhaps updating the vocabulary would make it more consistent.

**Comment by swcurran** [comment]
This looks very good and I think it should be accepted.

I do think the challenge will definitely be in the details and the evolution (interface, content) should occur as code, including documentation generated from code, and not via the HIPE process.  The concepts are excellent starting points, but I think developers, starting from here and driven by user stories, should define what's doable and what's the best way to achieve the vision.

**Comment by nage** [comment]
I have read the PR and am happy with the general proposals here.

File: text/0015-agent-test-suite-interface/README.md [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+    [Transport Adapters and `agtalk`](#transport-adapters-and-agtalk).)
+   * A backchannel way to ask the tested agent to initiate certain actions.
+   (We cannot see what a tested agent's outbound requests look like if all
+   it does is sit passively, making responses.) See 
+   [Asking for Action](#asking-for-action) below.
+   * A backchannel way to invoke setup and teardown logic, so tests can
+   satisfy deterministic pre- and post-conditions. We don't want later
+   tests to produce invalid results due to accumulated state from previous
+   tests. See [Setup and Teardown](#setup-and-teardown) below.
+   * A ledger that both it and the tested agent can refer to. This is
```
Comment (lovesh) [inline text/0015-agent-test-suite-interface/README.md]: What does this ledger record? Is there a separate backchannel or is there a message type for such communication?

File: text/0015-agent-test-suite-interface/README.md [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+others may use IRC. Since agent communication is transport agnostic, it
+is desirable to write tests once and then use them unmodified, over as
+many transports as are relevant.
+
+`agtest` will address this requirement by implementing an exceedingly
+simple transport protocol of its own--`agtalk`--and by providing
+pluggable adapters that recast the communication into the transport of
+interest.
+
+The agtalk transport works like this:
```
Comment (lovesh) [inline text/0015-agent-test-suite-interface/README.md]: Should agtalk guarantee delivery?

File: text/0015-agent-test-suite-interface/README.md [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+* To send a message, `agtest` will copy the bytes of the message into a
+new file in a folder on disk that has been designated the __outbox__.
```
Comment (lovesh) [inline text/0015-agent-test-suite-interface/README.md]: Is it mandatory to use files in this manner? Or should it be considered as a recommendation for building a "ordered, deliver only once" channel?

File: text/0015-agent-test-suite-interface/README.md [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+Multiple adapters will be bundled with `agtest`. Mostly, they should work
+out of the box. However, some agent authors may need to either tweak an
+adapter or write their own (for proprietary protocols or special
+environmental constraints).
+
+![agtest channels](agtest-channels.png)
+
+### Asking for Action
+
+When `agtest` wants to evaluate how an agent behaves as it proactively
```
Comment (lovesh) [inline text/0015-agent-test-suite-interface/README.md]: What are some examples of possible actions that can be taken on this backchannel? And what is an example of backchannel? Does communication on the backchannel follow `agtalk`?

File: text/0015-agent-test-suite-interface/README.md [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+makes requests of others, it will invoke the `agact` script. This script
+requires configuration or coding by the developers testing their agent;
+it cannot work out of the box, because test suite creators can't know
+the design parameters of the backchannel that `agact` uses. This is the
+blue "test active behavior" line running from `agtest` to `agent` in
+the diagram above.
```
Comment (lovesh) [inline text/0015-agent-test-suite-interface/README.md]: Is agact only concerned with the blue "test active behavior" line and the purple line is used by the agent being tested under normal (non-testing) circumstances? I think the answer is yes but confirming.

File: text/0015-agent-test-suite-interface/README.md [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+had `QueryInterface()`.
+ 
+An agent's interop profile can provide an objective, crisp way to answer
+interop questions. Such queries are encouraged as the interoperable way
+to communicate about interoperability.
+
+This leads to a definition of 2 canonical data formats. The first
+is the JSON representation of an interop profile. It looks like this:
+
+##### Interop Profile JSON
```
Comment (lovesh) [inline text/0015-agent-test-suite-interface/README.md]: Probably we can support 2 types of profiles, brief and detailed. Brief is sparse and does not contain divergent junctions or failed test names of interoperable or constrained junctions. Detailed contains all junctions and names of failing tests as well.

File: text/0015-agent-test-suite-interface/README.md [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+requested is, _Please tell me your interop profile for suite X_, where X is 
+identified by URI.
+
+Again, this is a fragment, allowing it to be put into an array of requests
+that are sent as a batch (and replied to as an array/batch as well).
+   
+# Reference
+[reference]: #reference
+
+A reference implementation of `agtest` is attached to this HIPE and
```
Comment (lovesh) [inline text/0015-agent-test-suite-interface/README.md]: Where is the implementation? The attached reference implementation of `atalk` is empty?

File: text/0015-agent-test-suite-interface/README.md [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+initiating action, or passively listening, or both? This embodies in
+agent space the human experience that many of us know, where it is
+possible to speak without listening, or to understand a language
+without being able to speak it. The other axis holds different __feature
+clusters__ -- which functional areas does the agent target? The set of
+feature clusters will likely grow over time.
+
+![axes of evaluation](matrix.png)
+
+Any given intersection in this matrix represents an __interop junction__.
```
Comment (lovesh) [inline text/0015-agent-test-suite-interface/README.md]: Does each test suite define its own threshold for the interop scores like 51% of tests should pass to be called `constrained` or these particular tests to pass to be called `constrained` or if these tests fail then certainly `divergent`

**Pull request merged**